File: app/src/main/java/com/arif/notifforwarder/AppFilterActivity.java
package com.arif.notifforwarder;

import android.content.SharedPreferences;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.BaseAdapter;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class AppFilterActivity extends AppCompatActivity {

    static class AppItem {
        String   name;
        String   packageName;
        Drawable icon;
    }

    private EditText  etSearch;
    private ListView  listView;
    private TextView  tvSelectedCount, tvFilterMode;
    private Button    btnSave;
    private Switch    switchFilterMode;

    private final List<AppItem> allApps  = new ArrayList<>();
    private final List<AppItem> filtered = new ArrayList<>();
    private final Set<String>   selected = new HashSet<>();
    private AppListAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_app_filter);

        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle("Filter Aplikasi");
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        etSearch        = findViewById(R.id.etSearch);
        listView        = findViewById(R.id.listView);
        tvSelectedCount = findViewById(R.id.tvSelectedCount);
        tvFilterMode    = findViewById(R.id.tvFilterMode);
        btnSave         = findViewById(R.id.btnSaveFilter);
        switchFilterMode = findViewById(R.id.switchFilterMode);

        SharedPreferences prefs = getSharedPreferences("notif_prefs", MODE_PRIVATE);
        Set<String> saved = prefs.getStringSet("allowed_packages", new HashSet<>());
        if (saved != null) selected.addAll(saved);
        switchFilterMode.setChecked(prefs.getBoolean("filter_enabled", false));

        updateModeLabel();
        switchFilterMode.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton btn, boolean checked) {
                updateModeLabel();
            }
        });

        adapter = new AppListAdapter();
        listView.setAdapter(adapter);
        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int pos, long id) {
                doToggle(pos, view);
            }
        });

        etSearch.addTextChangedListener(new TextWatcher() {
            @Override public void beforeTextChanged(CharSequence s, int a, int b, int c) {}
            @Override public void onTextChanged(CharSequence s, int a, int b, int c) {}
            @Override
            public void afterTextChanged(Editable s) { doFilter(s.toString()); }
        });

        btnSave.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) { doSave(); }
        });

        loadApps();
    }

    @Override
    public boolean onSupportNavigateUp() {
        finish();
        return true;
    }

    private void loadApps() {
        new Thread(new Runnable() {
            @Override
            public void run() {
                PackageManager pm = getPackageManager();
                List<ApplicationInfo> pkgs = pm.getInstalledApplications(PackageManager.GET_META_DATA);
                final List<AppItem> list = new ArrayList<>();
                for (ApplicationInfo info : pkgs) {
                    if ((info.flags & ApplicationInfo.FLAG_SYSTEM) != 0) continue;
                    AppItem item = new AppItem();
                    item.name        = pm.getApplicationLabel(info).toString();
                    item.packageName = info.packageName;
                    try { item.icon = pm.getApplicationIcon(info.packageName); }
                    catch (Exception ignored) {}
                    list.add(item);
                }
                Collections.sort(list, (a, b) -> a.name.compareToIgnoreCase(b.name));
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        allApps.clear(); allApps.addAll(list);
                        filtered.clear(); filtered.addAll(list);
                        adapter.notifyDataSetChanged();
                        updateCount();
                    }
                });
            }
        }).start();
    }

    private void doFilter(String q) {
        filtered.clear();
        String lower = q.toLowerCase();
        for (AppItem a : allApps) {
            if (q.isEmpty() || a.name.toLowerCase().contains(lower)
                    || a.packageName.toLowerCase().contains(lower)) {
                filtered.add(a);
            }
        }
        adapter.notifyDataSetChanged();
    }

    private void doToggle(int pos, View row) {
        AppItem app = filtered.get(pos);
        CheckBox cb = row.findViewById(R.id.checkboxApp);
        if (selected.contains(app.packageName)) {
            selected.remove(app.packageName);
            if (cb != null) cb.setChecked(false);
        } else {
            selected.add(app.packageName);
            if (cb != null) cb.setChecked(true);
        }
        updateCount();
    }

    private void doSave() {
        getSharedPreferences("notif_prefs", MODE_PRIVATE).edit()
                .putStringSet("allowed_packages", selected)
                .putBoolean("filter_enabled", switchFilterMode.isChecked())
                .apply();
        Toast.makeText(this, "Filter disimpan!", Toast.LENGTH_SHORT).show();
        finish();
    }

    private void updateCount() {
        tvSelectedCount.setText(selected.size() + " app dipilih");
    }

    private void updateModeLabel() {
        if (switchFilterMode.isChecked()) {
            tvFilterMode.setText("Mode: Hanya app yang dipilih");
            tvFilterMode.setTextColor(0xFF2E7D32);
        } else {
            tvFilterMode.setText("Mode: Semua app (filter nonaktif)");
            tvFilterMode.setTextColor(0xFF888888);
        }
    }

    class AppListAdapter extends BaseAdapter {
        @Override public int    getCount()              { return filtered.size(); }
        @Override public Object getItem(int pos)        { return filtered.get(pos); }
        @Override public long   getItemId(int pos)      { return pos; }

        @Override
        public View getView(int pos, View convertView, ViewGroup parent) {
            if (convertView == null) {
                convertView = LayoutInflater.from(AppFilterActivity.this)
                        .inflate(R.layout.item_app, parent, false);
            }
            AppItem app  = filtered.get(pos);
            ((ImageView) convertView.findViewById(R.id.ivAppIcon)).setImageDrawable(app.icon);
            ((TextView)  convertView.findViewById(R.id.tvAppName)).setText(app.name);
            ((TextView)  convertView.findViewById(R.id.tvPackageName)).setText(app.packageName);
            ((CheckBox)  convertView.findViewById(R.id.checkboxApp))
                    .setChecked(selected.contains(app.packageName));
            return convertView;
        }
    }
}

File: app/src/main/java/com/notifforwarder/app/AppFilterActivity.kt
package com.notifforwarder.app

import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

class AppFilterActivity : AppCompatActivity() {

    private lateinit var etSearch: EditText
    private lateinit var recyclerView: RecyclerView
    private lateinit var tvSelectedCount: TextView
    private lateinit var btnSave: Button
    private lateinit var switchFilterMode: Switch
    private lateinit var tvFilterMode: TextView

    private val allApps = mutableListOf<AppItem>()
    private val filteredApps = mutableListOf<AppItem>()
    private val selectedPackages = mutableSetOf<String>()
    private lateinit var adapter: AppAdapter

    data class AppItem(
        val name: String,
        val packageName: String,
        val icon: android.graphics.drawable.Drawable?
    )

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_app_filter)

        supportActionBar?.title = "Filter Aplikasi"
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        etSearch       = findViewById(R.id.etSearch)
        recyclerView   = findViewById(R.id.recyclerView)
        tvSelectedCount = findViewById(R.id.tvSelectedCount)
        btnSave        = findViewById(R.id.btnSaveFilter)
        switchFilterMode = findViewById(R.id.switchFilterMode)
        tvFilterMode   = findViewById(R.id.tvFilterMode)

        // Load saved state
        val prefs = getSharedPreferences("notif_prefs", MODE_PRIVATE)
        val saved = prefs.getStringSet("allowed_packages", emptySet()) ?: emptySet()
        val filterEnabled = prefs.getBoolean("filter_enabled", false)
        selectedPackages.addAll(saved)
        switchFilterMode.isChecked = filterEnabled

        updateFilterModeLabel()
        switchFilterMode.setOnCheckedChangeListener { _, _ -> updateFilterModeLabel() }

        // Setup RecyclerView
        adapter = AppAdapter(filteredApps, selectedPackages) { updateSelectedCount() }
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = adapter

        // Load installed apps di background
        loadInstalledApps()

        // Search
        etSearch.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable?) = filterList(s.toString())
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
        })

        btnSave.setOnClickListener {
            prefs.edit()
                .putStringSet("allowed_packages", selectedPackages)
                .putBoolean("filter_enabled", switchFilterMode.isChecked)
                .apply()
            Toast.makeText(this, "✅ Filter disimpan!", Toast.LENGTH_SHORT).show()
            finish()
        }
    }

    private fun loadInstalledApps() {
        Thread {
            val pm = packageManager
            val packages = pm.getInstalledApplications(PackageManager.GET_META_DATA)

            val apps = packages
                .filter { it.flags and ApplicationInfo.FLAG_SYSTEM == 0 } // hanya app user
                .map { info ->
                    AppItem(
                        name = pm.getApplicationLabel(info).toString(),
                        packageName = info.packageName,
                        icon = try { pm.getApplicationIcon(info.packageName) } catch (e: Exception) { null }
                    )
                }
                .sortedBy { it.name.lowercase() }

            runOnUiThread {
                allApps.clear()
                allApps.addAll(apps)
                filteredApps.clear()
                filteredApps.addAll(apps)
                adapter.notifyDataSetChanged()
                updateSelectedCount()
            }
        }.start()
    }

    private fun filterList(query: String) {
        filteredApps.clear()
        if (query.isEmpty()) {
            filteredApps.addAll(allApps)
        } else {
            filteredApps.addAll(allApps.filter {
                it.name.contains(query, ignoreCase = true) ||
                it.packageName.contains(query, ignoreCase = true)
            })
        }
        adapter.notifyDataSetChanged()
    }

    private fun updateSelectedCount() {
        tvSelectedCount.text = "${selectedPackages.size} app dipilih"
    }

    private fun updateFilterModeLabel() {
        if (switchFilterMode.isChecked) {
            tvFilterMode.text = "Mode: Hanya app yang dipilih ✅"
            tvFilterMode.setTextColor(getColor(android.R.color.holo_green_dark))
        } else {
            tvFilterMode.text = "Mode: Semua app (filter nonaktif)"
            tvFilterMode.setTextColor(getColor(android.R.color.darker_gray))
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        finish()
        return true
    }
}

// ── RecyclerView Adapter ─────────────────────────────────────────────────────

class AppAdapter(
    private val apps: List<AppFilterActivity.AppItem>,
    private val selected: MutableSet<String>,
    private val onChanged: () -> Unit
) : RecyclerView.Adapter<AppAdapter.ViewHolder>() {

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val ivIcon: ImageView   = view.findViewById(R.id.ivAppIcon)
        val tvName: TextView    = view.findViewById(R.id.tvAppName)
        val tvPkg: TextView     = view.findViewById(R.id.tvPackageName)
        val checkbox: CheckBox  = view.findViewById(R.id.checkboxApp)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_app, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val app = apps[position]
        holder.tvName.text = app.name
        holder.tvPkg.text = app.packageName
        holder.ivIcon.setImageDrawable(app.icon)
        holder.checkbox.isChecked = app.packageName in selected

        val toggle = {
            if (app.packageName in selected) {
                selected.remove(app.packageName)
                holder.checkbox.isChecked = false
            } else {
                selected.add(app.packageName)
                holder.checkbox.isChecked = true
            }
            onChanged()
        }

        holder.checkbox.setOnClickListener { toggle() }
        holder.itemView.setOnClickListener { toggle() }
    }

    override fun getItemCount() = apps.size
}
